tokenizer/scanner.py

```diff
@@ -1,3 +1,78 @@
 
-def scan(code):
-    print("scanner")
+def scan(filename):
+    print("scanner start")
+    token_specification = [
+        ('KEYWORD', ['if', 'then', 'else', 'while', 'do', 'end', 'procedure', 'var', 'begin', 'print']),
+        ('OPERATOR', ['=','+', '-', '*', '/', '==', '!=', '<=', '>=', '<', '>', 'and', 'or', 'not']),
+        ('SEPARATOR', [';']),
+        ('LPAR', ['(']),
+        ('RPAR', [')'])
+    ]
+    
+    tokens = []
+    
+    with open(filename, 'r', encoding='utf-8') as file:
+        for line in file:
+            input_string = line.strip()
+            i = 0
+            while i < len(input_string):
+                # ignore whitespace
+                if input_string[i].isspace():
+                    i += 1
+                    continue
+
+                matched = False
+
+                # Keyword
+                for token_type, values in token_specification:
+                    if token_type == 'KEYWORD':
+                        for keyword in values:
+                            if input_string[i:i+len(keyword)] == keyword and (i+len(keyword) == len(input_string) or not input_string[i+len(keyword)].isalnum()):
```

Comment:
What's the logic of the second and third condition?

Reply:
Check if it is the last word of the sentence. If it is, no additional check is needed. If it is not, check if any other word follows it. For example, 'while' is a keyword, but 'while1' is not.

Reply:
Hmm so when there is a special character in the string e.g. "while@", this will accept it right?